pkg/controllers/user/pipeline/hooks/drivers/github.go

```diff
@@ -0,0 +1,121 @@
+package drivers
+
+import (
+	"crypto/hmac"
+	"crypto/sha1"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"github.com/google/go-github/github"
+	"github.com/pkg/errors"
+	"github.com/rancher/rancher/pkg/controllers/user/pipeline/utils"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	"io/ioutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"net/http"
+	"strings"
+)
+
+const GithubWebhookHeader = "X-GitHub-Event"
+
+type GithubDriver struct {
+	Management *config.ManagementContext
+}
+
+func (g GithubDriver) Execute(req *http.Request) (int, error) {
+	var signature string
+	if signature = req.Header.Get("X-Hub-Signature"); len(signature) == 0 {
+		logrus.Errorf("receive github webhook,no signature")
+		return http.StatusUnprocessableEntity, errors.New("github webhook missing signature")
+	}
+	event := req.Header.Get(GithubWebhookHeader)
+	if event == "ping" {
+		return http.StatusOK, nil
+	} else if event != "push" {
+		//or "pull_request"
+		return http.StatusUnprocessableEntity, fmt.Errorf("not trigger for event:%s", event)
+	}
+
+	pipelineID := req.URL.Query().Get("pipelineId")
+	parts := strings.Split(pipelineID, ":")
+	if len(parts) < 0 {
+		return http.StatusUnprocessableEntity, fmt.Errorf("pipeline id '%s' is not valid", pipelineID)
+	}
+	ns := parts[0]
+	id := parts[1]
+	pipelines := g.Management.Management.Pipelines(ns)
+	pipelineExecutions := g.Management.Management.PipelineExecutions(ns)
```

Comment:
I would prefer that any API target used for update/create/remove/list, be passed as a registration, not formed on the fly based on g.management.management